maintenance_equipment_hierarchy/models/maintenance_equipment.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2019 Eficent Business and IT Consulting Services S.L.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class MaintenanceEquipment(models.Model):
+
+    _inherit = 'maintenance.equipment'
+    _parent_name = "parent_id"
+    _parent_store = True
+    _parent_order = 'name'
+
+    parent_id = fields.Many2one('maintenance.equipment', 'Parent Equipment',
+                                index=True, ondelete='cascade')
+    child_ids = fields.One2many('maintenance.equipment', 'parent_id',
+                                'Child Equipments')
+    parent_left = fields.Integer('Left Parent', index=1)
+    parent_right = fields.Integer('Right Parent', index=1)
+    child_count = fields.Integer(
+        compute='_compute_child_count',
+        string="Number of child equipments")
+
```

Comment:
```suggestion
    display_name = fields.Char(compute='_compute_display_name')

    @api.depends('name', 'parent_id.display_name')  # recursive definition
    def _compute_display_name(self):
        if self.parent_id:
            self.display_name = self.parent_id.display_name + ' / ' + self.name
        else:
            self.display_name = self.name

```

This is also needed IMO.

Reply:
I agree... In fact, if `display_name` field is defined as stored, it's possible to use it in a `_name_search()` implementation, as @etobella suggested.

And `_rec_name='display_name'` should be other improvement, as well

@astirpe what do you think?

Reply:
I don't like the idea to make display_name as stored. I also don't like the idea of  `_rec_name='display_name'`.

I would rather prefer a new dedicated field (eg.: `complete_name` or `full_path` or similar), computed and stored, to be used for the `_name_search()`; also the actual `_compute_display_name()` could benefit of it and be simplified.

Consider also to add field `parent_id` in the equipment listview.

Reply:
Thanks, I'll make the following changes as soon as possible:
- new `display_name` computed field
- `_name_get()` implementation
- new `complete_name` computed and stored field, which will be used by `_compute_display_name()` and `_name_get()` methods
- add `parent_id` to equipment list view.

Reply:
I've done the changes described above.
However, I didn't implement `_name_get()` because there was already a filter definition in the search view for equipment name, and I adapted that view code instead.
I've also added some demo data.
@astirpe could you check? Thanks!